### DISCUSSION TITLE
Unusable, blocked by Content Security Policy

### DISCUSSION THREAD

**Discussion by Grldk** [opcodesio/log-viewer#227] [Ideas]
This is probably more of a feature request than a bug, so I'm putting this here instead of in issues. Would like to use this log viewer, it looks great, but I cannot use it in production because it is blocked by our Content Security Policy. 

Our CSP forbids the use of `unsafe-eval`, and Vue by default uses `Function()`, which gets blocked by this CSP... There are [ways around](https://stackoverflow.com/questions/68459611/how-to-fix-unsafe-eval-error-with-vue3-for-the-client-side-version) this, but I'm not really familiar with Vue, so I'm afraid I can't be that much help to solve it.

**Comment by arukompas** [comment]
Hey @Grldk , thanks a lot for the report!

I'll have a look at it this weekend and hopefully there's an easy and non-breaking solution to this :) After all, assets are already pre-built and come part of the package, so there should be no need for `eval` use.

**Comment by Grldk** [comment]
Thanks for your quick reply!

It's an annoying consquence of using Vue.. And of using CSP in the first place, it's not widespread enough for people to take it into account unfortunately.. 

I've [tried](https://github.com/laravel/horizon/pull/1138) to fix it for Laravel Horizon, but failed (PR got reverted), and I'm not sure why. I don't really have any Vue experience, and not enough time or a use case to dive into it to see what the issue is.. Hope you can fix it for this package.. Thanks for your time!

**Comment by Grldk** [comment]
Hey @arukompas, had any luck with this?

**Comment by arukompas** [comment]
Hey @Grldk , I tried it for a while but I couldn't figure out how to make it CSP-friendly.

I have another idea, though, just not sure if it would work for your setup.

Would it be possible to override the CSP policy just for Log Viewer? Here's an example:

1. Publish the Log Viewer's layout - `php artisan vendor:publish --tag=log-viewer-views`
2. Open `resources/views/vendor/log-viewer/index.blade.php`
3. Add a meta tag to set a different CSP policy for this page
```diff
    <!-- ... -->
    <link rel="shortcut icon" href="{{ asset(mix('img/log-viewer-32.png', 'vendor/log-viewer')) }}">
+   <meta http-equiv="Content-Security-Policy" content="default-src 'self'; script-src 'self' 'unsafe-inline' 'unsafe-eval'; style-src 'self';">

    <title>Log Viewer{{ config('app.name') ? ' - ' . config('app.name') : '' }}</title>
    <!-- ... -->
```

This will allow unsafe-evals and, based on my testing, allows Log Viewer to work.

Let me know how it goes!

**Comment by Grldk** [comment]
Hey @arukompas, thanks again for your time.

Unfortunately that approach will not work. For CSP, [stricter rules take precedence.](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy#multiple_content_security_policies) We set those rules at the server level. So laxer rules set in PHP/Laravel (by middleware for example) or in the html do not work. 

If I find some time I might try to fix this for this package, but I'm quite short on time, and as I mentioned, earlier attempts to fix this for Laravel Horizon failed, so I'm not sure if I will be succesful..

**Reply by Grldk** [reply]
I have spent some time on this, have not been able to fix this, keep running into problems (some related to Vue, some to Webpack/mix) I don't understand or cannot figure out how to fix.. Might come back to this later, but for now decided to use something else

**Reply by arukompas** [reply]
I've also tried to change how Vue is bundled/built, but have been unsuccessful... I'm sorry @Grldk , but as of this moment I don't have any solution to this.

**Comment by hackerkok** [comment]
Hello!

I ran into the same issue. Is there any update on how to fix it? 

some [info](https://stackoverflow.com/questions/37155270/content-security-policy-csp-safe-usage-of-unsafe-eval)

**Reply by arukompas** [reply]
Unfortunately we don't have a solution yet. It's been a long time since we last looked, and not that many users seem affected by this. If you do have any hints or time to give it a try, that'd be most appreciated.

**Reply by jkokotinis** [reply]
fixed it and submitted a pull request.
thanks for the great plugin!

**Reply by arukompas** [reply]
@jkokotinis amazing, thank you again! New released tagged **v3.17.0** which now uses Vue runtime build.

@hackerkok @Grldk Please update to **v3.17.0**, should work in CSP environments now.
